Refactor docs / Maintenance of docs

ichard26 opened the pull request:
Resolves #1430.
Related #990 (please see #1457).
Resolves #1435.

**See the latest version of the refactored docs here:**
https://ichard26-testblackdocs.readthedocs.io/en/refactor_docs/index.html
(also is proof that this won't break the ReadTheDocs builds)

## Changes

### Refactor

- Split these sections from the README into their own files as per #1430:
  - `blackd` -> `docs/blackd.md`
  - `black-primer` -> `docs/black_primer.md`
  - `Editor integration` -> `docs/editor_integration.md`
  - `The Black code style` as per @zsol https://github.com/psf/black/issues/1430#issuecomment-630208621 -> `docs/the_black_code_style.md`
  - `Pragmatism` as per @zsol https://github.com/psf/black/issues/1430#issuecomment-631729659 -> `docs/the_black_code_style.md`

~~Yes, `Pragmatism` and `The Black code style` were merged into `docs/the_black_code_style_and_pragmatism.md`. **Please suggest a better title and filename for this poor piece of documentation.**~~ 

Thank you @hugovk for the suggestions! `the_black_code_style_and_pragmatism.md` has been renamed to `the_black_code_style.md`.

### Maintenance

- Fixed a few broken or likely to be broken links:
  - `README.md`:
    - `https://poetry.eustace.io/` -> `https://python-poetry.org/`
    - `https://docs.pipenv.org/` -> `https://readthedocs.org/projects/pipenv/`
  - `docs/editor_integration.md`
    - `https://github.com/proofit404/blacken` -> `https://github.com/pythonic-emacs/blacken`
- Deleted all of the symlinks in `docs/` since they serve no use now:
  - *you don't need me to list them... right?*
- Manually merged these changes since the docs that they were changing have been moved:
  - https://github.com/psf/black/commit/bb767863b0cc51897f9cd94da84b8b01f3936607
  - https://github.com/psf/black/commit/a2408b3cb23fe252b3674cee484d742496bb3411
  - https://github.com/psf/black/commit/397b16431e7e2af2869353d9310834f8c80e6cde
- Committed my changes from #1436.

### Features / Improvements

- ~~Made the sidebar navigation scrollable since we have so much packed into that sidebar that some of its contents are clipped. The only sad part of this change is that for users of Firefox, they will have to see an ugly scrollbar since Firefox disabled hiding scrollbar by default~~ (please see #1457)
- Completely reworked the logic in `docs/conf.py`.
  - Added a `DocSection` dataclass which represents sections of documentation
    - The `name` attribute is the section's name.
    - The `src` attribute is where to get the section's contents.
    - The `processors` attribute is processors to call on the section's contents before writing them back. This extensibility feature can be used to implement a link autoreplacer for example :)
    - The `out_filename` attribute is the filename to be used when writing the section. Only the merged `the_black_code_style_and_pragmatism` section uses this feature.
    - The `src_range` attribute is well... a `src_range` object
    - The `get_out_filename` method returns the `out_filename` attribute if set. If not, then it returns the section's name + `.md`
  - Added a `SrcRange` dataclass which represents a line range of the section's contents
  - Added a `get_contents` function which takes a `DocSection` and returns the contents of that section
  - Moved the header fixing code from the now deleted `generate_sections_from_readme` function into a processor which is ran for all sections generated from `README.md`.
  - Completely redone the `generate_sections_from_readme` function and renamed it to `get_sections_from_readme`. It now spits out a list of `DocSection` objects.
  - Created a new function called `process_sections` which takes `custom_sections` and `readme_sections` and processes them. The workflow of how it works is listed below:
    - Extend `custom_sections` with `readme_sections`
    - While iterating through each `DocSection` from the now extended `custom_sections`:
      - Check if a section which has the same name has been processed already. If so, then skip processing that section. This is how `custom_sections` overrides the default sections from `readme_sections`.
      - Get the target_path via `target_path: Path = CURRENT_DIR / section.get_out_filename()`
      - Get the section's contents using `get_contents(section)`
      - Run all processors defined from `section.processors` over the section's contents
      - Write the section's contents to `target_path`
      - Append the section's name to `processed_sections` so duplicates can be detected and skipped
- Using the new logic, I defined a `custom_sections` list which contains `DocSection` objects. This defines the custom / non-readme-based sections:
<details>

<summary>custom_sections</summary>

```py3
custom_sections = [
    DocSection("the_black_code_style", CURRENT_DIR / "the_black_code_style.md",),
    DocSection("pragmatism", CURRENT_DIR / "the_black_code_style.md",),
    DocSection("editor_integration", CURRENT_DIR / "editor_integration.md"),
    DocSection("blackd", CURRENT_DIR / "blackd.md"),
    DocSection("black_primer", CURRENT_DIR / "black_primer.md"),
    DocSection("contributing_to_black", CURRENT_DIR / ".." / "CONTRIBUTING.md"),
    DocSection("change_log", CURRENT_DIR / ".." / "CHANGES.md"),
]
```

</details>

hugovk commented on docs/black_primer.md:
````diff
@@ -0,0 +1,117 @@
+```text
+Usage: black-primer [OPTIONS]
+
+  primer - prime projects for blackening ... 🏴
````
From https://github.com/psf/black/commit/a2408b3cb23fe252b3674cee484d742496bb3411:
```suggestion
  primer - prime projects for blackening... 🏴
```

ichard26 replied:
I thought I manually merged that in, but it appears I didn't. Good catch!

hugovk commented on docs/black_primer.md:
```diff
@@ -0,0 +1,117 @@
+  -R, --rebase           Rebase project if already checked out  [default:
+                         False]
+
+  -w, --workdir PATH     Directory Path for repo checkouts  [default: /var/fol
```
```suggestion
  -w, --workdir PATH     Directory path for repo checkouts  [default: /var/fol
```

hugovk commented on docs/black_primer.md:
```diff
@@ -0,0 +1,117 @@
+      "cli_arguments": "List of extra CLI arguments to pass Black for this project",
+      "expect_formatting_changes": "Boolean to indicate that the version of Black is expected to cause changes",
+      "git_clone_url": "URL you would pass `git clone` to check out this repo",
+      "long_checkout": "Boolean to have repo skipped by defauult unless `--long-checkouts` is specified",
```
```suggestion
      "long_checkout": "Boolean to have repo skipped by default unless `--long-checkouts` is specified",
```

hugovk commented on docs/black_primer.md:
```diff
@@ -0,0 +1,117 @@
+                         ders/tc/hbwxh76j1hn6gqjd2n2sjn4j9k1glp/T/primer.20200
+                         517125229]
+
+  -W, --workers INTEGER  Number of parallel worker coroutines  [default: 69]
```
```suggestion
  -W, --workers INTEGER  Number of parallel worker coroutines  [default: 2]
```

hugovk commented on docs/conf.py:
```diff
@@ -14,28 +14,70 @@
+
+    Data:
+        name: The section's name. This will used to detect duplicate sections.
+        src: The filepath to get its contents
```
```suggestion
        src: The filepath to get its contents.
```

hugovk commented:
> Yes, `Pragmatism` and `The Black code style` were merged into `docs/the_black_code_style_and_pragmatism.md`. **Please suggest a better title and filename for this poor piece of documentation.**

Perhaps:

* The _Black_ code style
* `docs/style.md`

ichard26 commented:
Sorry that this is taking so long. I did a git merge master into refactor_docs and now I regret not doing  a rebase. This interactive rebase is not fun, full of merge conflicts!

ichard26 commented:
Alright, I've finished what I've wanted to do and resolved the many merge conflicts along the way. The key changes I made are the following:
* Make the ordering and default values of the attributes of the `DocSection` dataclass make a bit more sense (hopefully)
* Added logging so you know what the `process_sections` logic is doing
* Renamed `the_black_code_style_and_pragmatism.md` to `the_black_code_style`
  * and... renamed the nested `The Black code style` sub-header to just `Code style`
* Added @hugovk suggestions, thank you!

cooperlees commented on README.md:
```diff
@@ -1427,6 +538,7 @@ Multiple contributions by:
 - [Pablo Galindo](mailto:Pablogsal@gmail.com)
 - [Peter Bengtsson](mailto:mail@peterbe.com)
 - pmacosta
+- Richard Si
```
🤩 - Docs superstar
